src/h2_streams.erl

```diff
@@ -0,0 +1,541 @@
+-module(h2_streams).
+-include("http2.hrl").
+-include("h2_streams.hrl").
+
+-export(
+   [
+    new/1,
+    get/2,
+    close/4,
+    initiate_stream/7,
+    peer_initiated_stream/7,
+    replace/2,
+    send_what_we_can/4,
+    sort/1,
+    notify_pid/1,
+    update_all_recv_windows/2,
+    update_all_send_windows/2,
+    update_peer_max_active/2,
+    update_self_max_active/2
+   ]).
+
+-spec new(
+        client | server
+       ) -> streams().
+new(client) ->
+    #streams{type=client};
+new(server) ->
+    #streams{type=server}.
+
+-spec initiate_stream(
+        StreamId :: stream_id(),
+        NotifyPid :: pid(),
+        CBMod :: module(),
+        Socket :: sock:socket(),
+        InitialSendWindow :: pos_integer(),
+        InitialRecvWindow :: pos_integer(),
+        Streams :: streams()
+                   ) -> streams()
+                      | {error, term()} .
+initiate_stream(
+  StreamId,
+  NotifyPid,
+  CBMod,
+  Socket,
+  InitialSendWindow,
+  InitialRecvWindow,
+  Streams
+ ) ->
+    case
+        new_stream(
+          StreamId,
+          NotifyPid,
+          CBMod,
+          Socket,
+          InitialSendWindow,
+          InitialRecvWindow,
+          Streams#streams.self_initiated) of
+        {error, Error} ->
+            {error, Error};
+        NewStreamSet ->
+            Streams#streams{
+              self_initiated=NewStreamSet
+             }
+    end.
+
+-spec peer_initiated_stream(
+        StreamId :: stream_id(),
+        NotifyPid :: pid(),
+        CBMod :: module(),
+        Socket :: sock:socket(),
+        InitialSendWindow :: pos_integer(),
+        InitialRecvWindow :: pos_integer(),
+        Streams :: streams()
+                   ) -> streams()
+                      | {error, term()} .
+peer_initiated_stream(
+  StreamId,
+  NotifyPid,
+  CBMod,
+  Socket,
+  InitialSendWindow,
+  InitialRecvWindow,
+  Streams
+ ) ->
+    case
+        new_stream(
+          StreamId,
+          NotifyPid,
+          CBMod,
+          Socket,
+          InitialSendWindow,
+          InitialRecvWindow,
+          Streams#streams.peer_initiated) of
+        {error, Error} ->
+            {error, Error};
+        NewStreamSet ->
+            Streams#streams{
+              peer_initiated=NewStreamSet
+             }
+    end.
+
+new_stream(
+  StreamId,
+  NotifyPid,
+  CBMod,
+  Socket,
+  InitialSendWindow,
+  InitialRecvWindow,
+  StreamSet=#stream_set{max_active=unlimited}
+ ) ->
+    new_stream_(StreamId, NotifyPid, CBMod, Socket, InitialSendWindow, InitialRecvWindow, StreamSet);
```

Comment:
The function clause pattern matches on `max_active=unlimited` on line 109. 123 has no such guard, so both clauses are different reasons for why it's ok to make a new stream